### PR TITLE
ddl:check column count limit when add column

### PR DESCRIPTION
**Pull request by ciscoxll**
## What have you changed? (mandatory)
This pr fix problem, add a table column greater than `512` to report `errTooManyFields error`.

## What are the type of the changes (mandatory)?

Bug fix.

## How has this PR been tested (mandatory)?

Unit tests.

### CODE REVIEW

File: ddl/ddl_db_test.go [hunk 1]
```diff
@@ -1403,12 +1418,6 @@ func (s *testDBSuite) TestCreateTableTooLarge(c *C) {
 	}
 	sql += ");"
 	s.testErrorCode(c, sql, tmysql.ErrTooManyFields)
-
-	originLimit := ddl.TableColumnCountLimit
```
Comment (shenli) [inline ddl/ddl_db_test.go]: Why remove this?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -539,6 +539,13 @@ func checkTooManyColumns(colDefs []*ast.ColumnDef) error {
 	return nil
 }
 
+func checkAddColumnTooManyColumns(oldCols int, newColDefs []*ast.ColumnDef) error {
+	if colLen := oldCols + len(newColDefs); colLen > TableColumnCountLimit {
```
Comment (shenli) [inline ddl/ddl_api.go]: if oldCols + len(newColDefs) > TableColumnCountLimit

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1120,6 +1127,10 @@ func (d *ddl) AddColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTab
 		return errors.Trace(infoschema.ErrTableNotExists.GenByArgs(ti.Schema, ti.Name))
 	}
 
+	if err = checkAddColumnTooManyColumns(len(t.Meta().Columns), spec.NewColumns); err != nil {
```
Comment (winkyao) [inline ddl/ddl_api.go]: we need to check the columns count in onAddColumn stage. Imagine parallel adding column, can lead to TiDB exceed the limit.

File: ddl/ddl_db_test.go [hunk 4]
```diff
@@ -887,6 +887,21 @@ func (s *testDBSuite) TestColumn(c *C) {
 	s.tk.MustExec("drop table t2")
 }
 
+func (s *testDBSuite) TestAddColumnTooLarge(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test")
+	s.tk.MustExec("create table t_column_too_large(abc int)")
+	count := 511
```
Comment (winkyao) [inline ddl/ddl_db_test.go]: Do not hard code 511, use variable `TableColumnCountLimit `

File: ddl/column.go [hunk 5]
```diff
@@ -135,6 +135,11 @@ func (d *ddl) onAddColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		return ver, errors.Trace(err)
 	}
 
+	if err = checkAddColumnTooManyColumns(len(tblInfo.Columns)); err != nil {
```
Comment (zhexuany) [inline ddl/column.go]: Just wondered， is it ok if we can move this snippet up to line 129? Per my naive understanding, `tblInfo` is ready to be used at line 129, hence validating check can be done at then.

**Comment by ciscoxll** [comment]
@shenli @zhexuany  PTAL

**Comment by zhexuany** [comment]
@ciscoxll please fix CI.

File: ddl/ddl_db_test.go [hunk 6]
```diff
@@ -923,6 +923,21 @@ func (s *testDBSuite) TestColumn(c *C) {
 	s.tk.MustExec("drop table t2")
 }
 
+func (s *testDBSuite) TestAddColumnTooLarge(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test")
+	s.tk.MustExec("create table t_column_too_large(abc int)")
```
Comment (zimulala) [inline ddl/ddl_db_test.go]: s/too_large/too_many

File: ddl/column.go [hunk 7]
```diff
@@ -125,7 +125,10 @@ func (d *ddl) onAddColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	// if errorBeforeDecodeArgs {
 	// 	return ver, errors.New("occur an error before decode args")
 	// }
-
+	if err = checkAddColumnTooManyColumns(len(tblInfo.Columns)); err != nil {
```
Comment (zimulala) [inline ddl/column.go]: I think `tblInfo.Columns` doesn't the old columns when the new column's state isn't `StateNone`.

Comment (ciscoxll) [reply ddl/column.go]: @zimulala execute the `AddColumn` once and add only one column, so the old columns plus 1 is the new column's state.

File: ddl/column.go [hunk 8]
```diff
@@ -125,7 +125,10 @@ func (d *ddl) onAddColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	// if errorBeforeDecodeArgs {
 	// 	return ver, errors.New("occur an error before decode args")
 	// }
-
+	if err = checkAddColumnTooManyColumns(tblInfo.Columns); err != nil {
```
Comment (zimulala) [inline ddl/column.go]: I don't understand. `onAddColumn` will be called 4 times.
My opinion is that `tblInfo.Columns` contains `tblInfo.Columns` and the new column when the new column's state isn't StateNone.
e.g.
tblInfo.Columns is c1, the new column is c2.
At first, c2 is StateNone, so tblInfo.Columns is only c1.
Next, the c2 is StateDeleteOnly, so tblInfo.Columns are c1 and c2.

Comment (ciscoxll) [reply ddl/column.go]: I understand what you mean.

**Pull request merged**